Jeff/step 20

jeff-yoo-rakuma opened the pull request:
# Overview
This PR includes changes for adding labels to tasks.

# Changes
## Backend
* Add new model Label
* Add many to many association between Task and Label
* Change Task endpoint payload to include labels
* Search for label names as well as task names when `search` param is passed for GET tasks
* Add relevant tests
## Frontend
* Add option to add labels when creating/editing tasks
* Add labels to each task when displaying

# Tests
## task-service
```
root@a206bd3b07af:/app# rspec
.....................................................................

Finished in 4.09 seconds (files took 7.78 seconds to load)
69 examples, 0 failures
```
## user-service
```
root@0253241df354:/app# rspec
...................................

Finished in 2.56 seconds (files took 7.92 seconds to load)
35 examples, 0 failures
```

jeff-yoo-rakuma committed:
Add new model, Label

jeff-yoo-rakuma committed:
Make label searchable with search param on task GET

jeff-yoo-rakuma committed:
Add label to frontend UI

jeff-yoo-rakuma committed:
Update logical relation diagram

yo-it-engineer commented on todos-app/backend/task_service/spec/routing/labels_routing_spec.rb:
```diff
@@ -0,0 +1,31 @@
+      expect(patch: '/labels/1').to route_to('labels#update', id: '1')
+    end
+
+    it 'routes to #destroy' do
+      expect(delete: '/labels/1').to route_to('labels#destroy', id: '1')
+    end
```
It's my first time seeing a routing_spec. Where do you learn about this?

https://relishapp.com/rspec/rspec-rails/docs/routing-specs
> Simple apps with nothing but standard RESTful routes won't get much value from
routing specs, but they can provide significant value when used to specify
customized routes, like vanity links, slugs, etc.

jeff-yoo-rakuma replied:
It's actually created automatically by `rails g` 😅
I guess it's better than not having one?

yo-it-engineer replied:
I see!
I thought you added it yourself. If it was auto generated, then just leave it there.

> I guess it's better than not having one?

Yeah, that's right.

yo-it-engineer commented on todos-app/frontend/src/components/TaskCreateModal.vue:
```diff
@@ -32,6 +32,17 @@
 
             <label class="form-label mt-4">Due Date</label>
             <input type="datetime-local" class="form-control" v-model="task.due_datetime">
+
+            <label class="form-label mt-4">Tags</label>
+            <div class="d-flex flex-wrap">
+              <span v-for="(label, i) in task.labels" :key="i" class="badge rounded-pill badge-outline-secondary m-1">#{{label}}<i type="button" class="bi bi-x" @click="removeLabel(label)"></i></span>
+            </div>
+            <div class="input-group">
+              <input name="label" type="text" class="form-control" placeholder="Max 15 characters without space" v-model="labelInput">
+              <button type="button" class="btn btn-outline-secondary input-group-text" @click="addLabel">
+                <i class="bi bi-plus-circle"></i>
+              </button>
+            </div>
```
![image](https://user-images.githubusercontent.com/10822260/149431158-7cefda9f-eae0-42b5-b866-b3f5ecb8c2c7.png)

Wow. You are good at creating frontend UI. How long does it take you to come out with this part?

jeff-yoo-rakuma replied:
I got the idea from Vuetify's chip (https://vuetifyjs.com/en/components/chips/), but I couldn't use Vuetify in this project because it's not compatible with Vue 3 yet. I wanted to try out Vue 3 so I had to go with vanilla Bootstrap this time 🥲
I played around with css a bit to copy the looks 😁

yo-it-engineer commented on todos-app/frontend/src/components/TaskEditModal.vue:
```diff
@@ -32,6 +32,17 @@
             <label class="form-label mt-4">Due Date</label>
             <input name="due_datetime" type="datetime-local" class="form-control" :value="task.due_datetime ? task.due_datetime.substring(0, 16) : null">
+
+            <label class="form-label mt-4">Tags</label>
+            <div class="d-flex flex-wrap">
+              <span v-for="(label, i) in labels" :key="i" class="badge rounded-pill badge-outline-secondary m-1">#{{label}}<i type="button" class="bi bi-x" @click="removeLabel(label)"></i></span>
+            </div>
+            <div class="input-group">
+              <input name="label" type="text" class="form-control" placeholder="Max 15 characters without space" v-model="labelInput">
+              <button type="button" class="btn btn-outline-secondary input-group-text" @click="addLabel">
+                <i class="bi bi-plus-circle"></i>
+              </button>
+            </div>
```
It seemed like this part is kinda same with the TaskCreateModal.vue part.
Do vue file have something that can group and re-use components like `partial` in rails?(Usually for rails, we will use partial for the whole form and reuse it in `create` and `edit`'s html

jeff-yoo-rakuma replied:
That's probably a wiser choice. I just didn't get to refactor it 🙃

yo-it-engineer reviewed (commented):
All the requirements was met! Good work!

There are two points that have my attention:

1. Not a big issue, but there seemed to be having a validation problem when task is inputted but not pressing the `+` button to make it a tag. refer video:

https://user-images.githubusercontent.com/10822260/149431911-f6c50a15-6ddd-47d7-895f-1affb8e019f4.mov

Maybe we should ignore the inputed text if the user is not pressing the `+` button. Or we should let JS do the work and run the `+` button's action too.(a bit troublesome 😄 )

Anyway, it doesn't have big impact to the app so It's fine too the way it is.

2. You keep the label record in the DB even when user remove it from their task. Do you have any intention for leaving it like that? It will result in many unused labels in the DB that is not linked to any task.

jeff-yoo-rakuma committed:
Removed empty array validation

jeff-yoo-rakuma committed:
Add endpoint for deleting a user's labels

jeff-yoo-rakuma committed:
Add request for removing labels when deleting a user

jeff-yoo-rakuma commented:
@yo-it-engineer 

>Not a big issue, but there seemed to be having a validation problem when task is inputted but not pressing the + button to make it a tag. refer video:

It's because I put this validation on the backend and the client was sending an empty array for labels.
```ruby
    unless params[:task][:labels].is_a?(Array) && !params[:task][:labels].empty?
      raise Exceptions::InvalidLabelParams, 'Labels attribute must be a non-empty array'
    end
```
Now you mentioned it, I don't think this makes much sense. I changed it so that it can accept an empty array.
Thank you for spotting this!

>You keep the label record in the DB even when user remove it from their task. Do you have any intention for leaving it like that? It will result in many unused labels in the DB that is not linked to any task.

That's very true. I thought I added that part, but turns out I didn't 🤦 Added it now!

yo-it-engineer reviewed (approved):
I had confirmed the changes! LGTM!

![LGTM](https://image.lgtmoon.dev/85891)
